.github/workflows/tests.yml

```diff
@@ -150,16 +150,42 @@ jobs:
     # Note: sqlite only; no postgres
     if: ${{ !cancelled() && !failure() }} # Allow previous steps to be skipped, but not fail
     needs: linting-done
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-20.04
     steps:
       - uses: actions/checkout@v2
-      - name: Test with old deps
-        uses: docker://ubuntu:focal # For old python and sqlite
-        # Note: focal seems to be using 3.8, but the oldest is 3.7?
-        # See https://github.com/matrix-org/synapse/issues/12343
+
+      # There aren't wheels for some of the older deps, so we need to install
+      # their build dependencies
+      - run: |
+          sudo apt-get -qq install build-essential libffi-dev python-dev \
+          libxml2-dev libxslt-dev xmlsec1 zlib1g-dev libjpeg-dev libwebp-dev
+
+      - uses: actions/setup-python@v4
+        with:
+          python-version: '3.7'
+
+      # Calculating the old-deps actually takes a bunch of time, so we cache the
+      # pyproject.toml / poetry.lock.
+      - uses: actions/cache@v3
+        id: cache-poetry-old-deps
+        name: Cache poetry.lock
         with:
-          workdir: /github/workspace
-          entrypoint: .ci/scripts/test_old_deps.sh
+          path: |
+            poetry.lock
+            pyproject.toml
```

Comment:
I found this confusing; maybe worth a comment noting that the reason we cache pyproject.toml is because the step beneath patches it.

Reply:
My understanding:

- We compute a cache key based on the pyproject.toml in the checkout
- If there's a hit, we retrieve the _patched_ pyproject.toml and poetry.lock from the cache
- Otherwise we patch pyproject.toml and regenerate the lockfile
- During GHA cleanup we will write the patched pyproject.toml and poetry.lock to the cache for future runs

It does need a little bit of thought though: a quick comment SoundsGTM